interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,405 @@
+	id := atomic.AddUint64((*uint64)(unsafe.Pointer(&pdb.maxIDMmap[0])), 1)
+	idStr := fmt.Sprintf("%016X", id)
+	return idStr
+}
+
+// AddOrMerge checks if the given prompt contents are identical to an existing
+// prompt and, if so, merges with it by adding the given listenerReq to it.
+// Otherwise, adds a new prompt with the given contents to the prompt DB.
+// If an error occurs, no change is made to the DB.
+//
+// If the prompt was merged with an identical existing prompt, returns the
+// existing prompt and true, indicating it was merged. If a new prompt was
+// added, returns the new prompt and false, indicating the prompt was not
+// merged.
+//
+// The caller must ensure that the given permissions are in the order in which
+// they appear in the available permissions list for the given interface.
+func (pdb *PromptDB) AddOrMerge(metadata *prompting.Metadata, path string, permissions []string, listenerReq *listener.Request) (*Prompt, bool, error) {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, ok := pdb.perUser[metadata.User]
+	if !ok {
+		pdb.perUser[metadata.User] = &userPromptDB{
+			ByID: make(map[string]*Prompt),
+		}
+		userEntry = pdb.perUser[metadata.User]
+	}
+
+	availablePermissions, err := prompting.AvailablePermissions(metadata.Interface)
+	if err != nil {
+		// Error should be impossible, since caller has already validated that
+		// iface is valid, and tests check that all valid interfaces have valid
+		// available permissions returned by AvailablePermissions.
+		return nil, false, err
+	}
+
+	constraints := &PromptConstraints{
+		Path:                 path,
+		Permissions:          permissions,
+		AvailablePermissions: availablePermissions,
+	}
+
+	// Search for an identical existing prompt, merge if found
+	for _, prompt := range userEntry.ByID {
+		if prompt.Snap == metadata.Snap && prompt.Interface == metadata.Interface && prompt.Constraints.equals(constraints) {
+			prompt.listenerReqs = append(prompt.listenerReqs, listenerReq)
+			// Although the prompt itself has not changed, re-record a notice
+			// to re-notify clients to respond to this request. A client may
+			// have replied with a malformed response and not retried after
+			// receiving the error, so this notice encourages it to try again
+			// if the user retries the operation.
+			pdb.notifyPrompt(metadata.User, prompt.ID, nil)
+			return prompt, true, nil
+		}
+	}
+
+	if len(userEntry.ByID) >= maxOutstandingPromptsPerUser {
+		// Too many outstanding prompts, auto-deny this one
+		sendReply(listenerReq, false)
+		return nil, false, ErrTooManyPrompts
+	}
+
+	id := pdb.nextID()
+	timestamp := time.Now()
+	prompt := &Prompt{
+		ID:           id,
+		Timestamp:    timestamp,
+		Snap:         metadata.Snap,
+		Interface:    metadata.Interface,
+		Constraints:  constraints,
+		listenerReqs: []*listener.Request{listenerReq},
+	}
+	userEntry.ByID[id] = prompt
+	pdb.notifyPrompt(metadata.User, id, nil)
```

Comment:
I think the only remaining issue I have now is that the notify code is called with the lock held and we can't foresee if the callee won't block for any reason. Since we already have the user ID and prompt ID, this call could be done after we release the db lock?

Reply:
I would really prefer to record the notice while the requestprompts lock is held, since this guarantees the order in which the notices are issued. This lets us rely nicely on the fact that the notice when a prompt is resolved will always be the final notice for that prompt, and thus the notice data indicating the resolved reason will be preserved. We don't need to guarantee this, but it would be much nicer for clients.

Discussed more here: https://warthogs.atlassian.net/browse/SNAPDENG-15165

Reply:
not sure it should block landing this but we have the issues that emitting notices atm always require the state lock so we will hold this and then need to get the state lock to emit the notice. Maybe there's a way out once we undetstand more what should happens on snapd restarts. We already know that it doesn't make sense to keep prompt notices across reboots

Reply:
Right, I'm not sure what we can do to improve this at the moment. Perhaps in an ideal world notices would be decoupled from the state lock entirely, so they can relay information while the lock is held. And so that recording notices doesn't lock up the state all the time.

Perhaps something can be optimized so that the `notifyPrompt` closure kicks off the notice recording, guaranteeing order, but doesn't wait for the state lock to be acquired and released.

In any case, I think we want to call `notifyPrompt` with the prompt DB lock held, and it's up to the closure implementation to guarantee that it doesn't block.